modules/programs/nix-your-shell.nix

```diff
@@ -0,0 +1,21 @@
+{ config, lib, pkgs, ... }:
+let
+  inherit (lib) genAttrs getExe;
+  cfg = config.programs.nix-your-shell;
+
+  # In principle `bash` is supported too, but...  😹
+  shells = [ "fish" "ion" "nushell" "zsh" ];
+  programs = [ "nix" "nix-shell" ];
+in {
+  meta.maintainers = with lib.maintainers; [ nicoo ];
+
+  options.programs.nix-your-shell.enable = lib.mkEnableOption ''
+    `nix-your-shell`, a wrapper for `nix develop` or `nix-shell`
+    to run the same shell inside the new environment.
+  '';
+
+  config.programs = lib.mkIf cfg.enable (genAttrs shells (shell: {
+    shellAliases = genAttrs programs
+      (program: "${getExe pkgs.nix-your-shell} ${shell} ${program} --");
```

Comment:
I'm not certain it is always the best idea to just make a plain alias. Consider, for example, fish. We have

```
$ nix-your-shell fish
# If you see this output, you probably forgot to pipe it into `source`:
# nix-your-shell fish | source

function nix-shell --description "Start an interactive shell based on a Nix expression"
    nix-your-shell fish nix-shell -- $argv
end

function nix --description "Reproducible and declarative configuration management"
    nix-your-shell fish nix -- $argv
end
```

So it will also provide a nice command description, which will be missing in the plain alias definition.

For the nushell alias, I'm not certain it will work at all. My understanding is that nushell aliases cannot take command line parameters…

Reply:
> on ZSH it also breaks autocompletion

I usually use `compdef` for this:

```zsh
compdef _actual_command wrapped_command
```

Reply:
Yea, we should really use `nix-your-shell fish | source`, like it is done in `zoxide`, `starship`, etc.